src/modules/logicModule/paymentClient/ERC20PaymentClient.sol

```diff
@@ -208,4 +184,53 @@ abstract contract ERC20PaymentClient is
             revert Module__ERC20PaymentClient__InvalidRecipient();
         }
     }
+
+    //--------------------------------------------------------------------------
+    // {ERC20PaymentClient} Function Implementations
+
+    /// @dev Ensures `amount` of payment tokens exist in address(this).
+    function _ensureTokenBalance(uint amount) internal virtual {
+        uint balance = __Module_orchestrator.token().balanceOf(address(this));
+
+        if (balance < amount) {
+            // Trigger callback from orchestrator to transfer tokens
+            // to address(this).
+            bool ok;
+            (ok, /*returnData*/ ) = __Module_orchestrator.executeTxFromModule(
+                address(__Module_orchestrator.fundingManager()),
+                abi.encodeWithSignature(
+                    "transferOrchestratorToken(address,uint256)",
+                    address(this),
+                    amount - balance
+                )
+            );
+
+            if (!ok) {
+                revert Module__ERC20PaymentClient__TokenTransferFailed();
+            }
+        }
+    }
+
+    /// @dev Ensures `amount` of token allowance for payment processor(s).
+    function _ensureTokenAllowance(IPaymentProcessor spender, uint amount)
+        internal
+        virtual
+    {
+        IERC20 token = __Module_orchestrator.token();
+        uint allowance = token.allowance(address(this), address(spender));
+
+        if (allowance < amount) {
+            token.safeIncreaseAllowance(address(spender), amount - allowance);
+        }
```

Comment:
This opens up another can of worms. Maybe something that we should decide as a team over a call.

So, the question is, do you want to also *decrease* the allowance of the *spender* if it is more than the `amount` ?

I'm asking this because the way the contracts are right now, depending on the front-end applications built on top of it, the users could be asked to give infinite approval to some contracts.

And if that is the case, there's this huge debate around safety vs convenience regarding infinite approvals. [This blog](https://hackernoon.com/erc20-infinite-approval-a-battle-between-convenience-and-security-lk60350r) encapsulates the essence of the argument.

Reply:
I think this should be brought up in a new issue instead of here, because this was more of an refactoring PR than anything else.
I'll create one for it.

Reply:
But isnt this purely restricted to paymentProcessor that is connected to its own payment Processor?
I dont really see the problem here

Reply:
Yeah, my comment was on the broader theme, but I think your point is valid and we can ignore that comment.